src/System.Memory/src/System/MemoryExtensions.cs

```diff
@@ -10,13 +10,317 @@
+                end--;
+            }
+            return span.Slice(0, end + 1);
+        }
+
+        /// <summary>
+        /// Removes all leading and trailing occurrences of a set of characters specified 
+        /// in a readonly span from the span.
+        /// </summary>
+        /// <param name="span">The source span from which the characters are removed.</param>
+        /// <param name="trimChars">The span which contains the set of characters to remove.</param>
+        public static ReadOnlySpan<char> Trim(this ReadOnlySpan<char> span, ReadOnlySpan<char> trimChars)
+        {
+            if (span.Length == 0)
+                return default;
+            if (trimChars.Length == 0)
+                return span;
+
+            return span.TrimStart(trimChars).TrimEnd(trimChars);
+        }
+
+        /// <summary>
+        /// Removes all leading occurrences of a set of characters specified 
+        /// in a readonly span from the span.
+        /// </summary>
+        /// <param name="span">The source span from which the characters are removed.</param>
+        /// <param name="trimChars">The span which contains the set of characters to remove.</param>
+        public static ReadOnlySpan<char> TrimStart(this ReadOnlySpan<char> span, ReadOnlySpan<char> trimChars)
+        {
+            if (span.Length == 0)
+                return default;
+            if (trimChars.Length == 0)
+                return span;
+
+            ref char first = ref MemoryMarshal.GetReference(span);
+            ref char firstTrimChar = ref MemoryMarshal.GetReference(trimChars);
+            int start = 0;
+#if !netstandard11
+            if (Vector.IsHardwareAccelerated && span.Length >= 2 * Vector<ushort>.Count)
+            {
+                var equalityTester = new Vector<ushort>(ushort.MaxValue);
+                while (start < span.Length - Vector<ushort>.Count)
+                {
+                    Vector<ushort> value = Unsafe.ReadUnaligned<Vector<ushort>>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref first, start)));
+                    Vector<ushort> comparison = Vector<ushort>.Zero;
+                    int j = 0;
+                    for (; j < trimChars.Length; j++)
+                    {
+                        var mask = new Vector<ushort>(Unsafe.Add(ref firstTrimChar, j));
+                        comparison |= Vector.Equals(value, mask);
+                    }
+                    if (!equalityTester.Equals(comparison))
+                        break;
+                    start += Vector<ushort>.Count;
+                }
+            }
+#endif
+            while (start < span.Length)
+            {
+                int j = 0;
+                while (j < trimChars.Length)
+                {
+                    if (Unsafe.Add(ref first, start) == Unsafe.Add(ref firstTrimChar, j))
+                        break;
+                    j++;
+                }
+                if (j == trimChars.Length)
+                    break;
```

Comment:
maybe we can Slice and return from here. and at the end we can just return the original span without slicing

Reply:
that won't always work. we have other breaks where we would need to slice as well.

Reply:
you are right. I thought we are trimming a string and not a set of characters. this fine then.